lib/diskoverflow/sorted.go

```diff
@@ -0,0 +1,274 @@
+// Copyright (C) 2018 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package diskoverflow
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"sort"
+)
+
+const suffixLength = 8
+
+// Sorted is a disk-spilling container, that sorts added values by the
+// accompanying keys.
+// You must not add new values after calling PopFirst/-Last.
+type Sorted interface {
+	Add(k []byte, v Value)
+	PopFirst(v Value) bool
+	PopLast(v Value) bool
+	NewIterator() Iterator
+	NewReverseIterator() Iterator
+	Bytes() int
+	Items() int
+	SetOverflowBytes(bytes int)
+	Close()
+}
+
+type sorted struct {
+	commonSorted
+	base
+}
+
+type keyIterator interface {
+	Iterator
+	key() []byte
+}
+
+type commonSorted interface {
+	common
+	add(k []byte, v Value)
+	PopFirst(v Value) bool
+	PopLast(v Value) bool
+	newIterator(p iteratorParent, reverse bool) keyIterator
+}
+
+// NewSorted returns an implementaiton of Sorted, spilling to disk at location.
+func NewSorted(location string) Sorted {
+	o := &sorted{base: newBase(location)}
+	o.commonSorted = &memorySorted{}
+	return o
+}
+
+func (o *sorted) Add(k []byte, v Value) {
+	if o.startSpilling(o.Bytes() + v.ProtoSize()) {
+		d, err := v.Marshal()
+		errPanic(err)
+		newSorted := newDiskSorted(o.location)
+		it := o.NewIterator().(keyIterator)
+		for it.Next() {
+			v.Reset()
+			it.Value(v)
+			newSorted.add(it.key(), v)
+		}
+		it.Release()
+		o.commonSorted.Close()
+		o.commonSorted = newSorted
+		o.spilling = true
+		v.Reset()
+		errPanic(v.Unmarshal(d))
+	}
+	o.add(k, v)
+}
+
+func (o *sorted) NewIterator() Iterator {
+	return o.newIterator(false)
+}
+
+func (o *sorted) NewReverseIterator() Iterator {
+	return o.newIterator(true)
+}
+
+func (o *sorted) newIterator(reverse bool) Iterator {
+	if o.iterating {
+		panic(concurrencyMsg)
+	}
+	o.iterating = true
+	return o.commonSorted.newIterator(o, reverse)
+}
+
+// Close is just here to catch deferred calls to Close, such that the correct
+// method is called in case spilling happened.
+func (o *sorted) Close() {
+	o.commonSorted.Close()
+}
+
+func (o *sorted) String() string {
+	return fmt.Sprintf("Sorted@%p", o)
+}
+
+func (o *sorted) released() {
+	o.iterating = false
+}
+
+// memorySorted is basically a slice that keeps track of its size and supports
+// sorted iteration of its element.
+type memorySorted struct {
+	bytes    int
+	outgoing bool
+	values   []Value
+	keys     [][]byte
+}
+
+func (o *memorySorted) add(k []byte, v Value) {
+	if o.outgoing {
+		panic("Add/Append may never be called after PopFirst/PopLast")
+	}
+	o.values = append(o.values, v)
+	o.keys = append(o.keys, k)
```

Comment:
This should use a heap.

Reply:
So I just read up on heaps (well, a bit): That means only one direction, i.e. one of `GetFirst`/`GetLast` is possible, and that has to be determined at object creation time - correct?